src/canisters/registrar/src/service.rs

```diff
@@ -1288,6 +1335,139 @@ impl RegisterCoreContext {
     }
 }
 
+pub struct RegistrationNameQueryContext<'a> {
+    token_index_store: &'a TokenIndexStore,
+    registration_store: &'a RegistrationStore,
+}
+
+impl<'a> RegistrationNameQueryContext<'a> {
+    pub fn new(
+        token_index_store: &'a TokenIndexStore,
+        registration_store: &'a RegistrationStore,
+    ) -> Self {
+        Self {
+            token_index_store: token_index_store,
+            registration_store: registration_store,
+        }
+    }
+
+    pub fn get_all_unexpired_registrations(&self, now: u64) -> Vec<UnexpiredRegistrationAggDto> {
+        let registration_names = self.token_index_store.get_registrations();
+        let mut valid_registration_names = Vec::new();
+
+        for registration_name_ref in registration_names {
+            let registration_name = registration_name_ref.borrow();
+            let id = encode_token_id(
+                common::token_identifier::CanisterId(get_named_get_canister_id(
+                    CanisterNames::Registrar,
+                )),
+                registration_name.get_index(),
+            );
+            let registration_result =
+                self.get_unexpired_registration_by_name(&registration_name.get_name(), now);
+            match registration_result {
+                Ok(registration) => {
```

Comment:
If there is no data, then the previous step of constructing the id is invalid and the logic should be moved here